src/libos/crates/async-rt/src/time/timeout.rs

```diff
@@ -0,0 +1,54 @@
+use pin_project::pin_project;
+
+use crate::prelude::*;
+use crate::time::timer::TimerEntry;
+
+/// Timeout is a decorator future that can wrap any future to make a new future
+/// that completes until the internal future completes or the timeout expires.
+///
+/// Since it's a future, it doesn't execute unless you `.await` or `poll`.
+#[pin_project]
+pub struct Timeout<'a, F> {
+    #[pin]
+    future: F,
+    #[pin]
+    timer: TimerEntry,
+
+    timeout: &'a mut Duration,
+}
+
+impl<'a, F> Timeout<'a, F> {
+    pub fn new(future: F, timeout: &'a mut Duration) -> Self {
+        Self {
+            future,
+            timer: TimerEntry::new(*timeout),
+            timeout,
+        }
+    }
+}
+
+impl<F: Future> Future for Timeout<'_, F> {
+    type Output = Result<F::Output>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        // First, we poll the internel future.
+        // If the future is ready, update the remaining duration and return Ready.
+        if let Poll::Ready(output) = this.future.poll(cx) {
+            **this.timeout = this.timer.remain_duration();
+            return Poll::Ready(Ok(output));
+        }
+
+        // Second, we poll the timer.
+        // If the timer is ready, the timeout has expired, return Ready.
+        // else return Pending to wait next poll.
+        match this.timer.poll(cx) {
+            Poll::Ready(_) => {
+                **this.timeout = Duration::new(0, 0);
```

Comment:
Use `Duration::ZERO` instead.

Reply:
I tried it before, `ZERO` is not supported by our env's rust version.